### ISSUE TITLE
cargo forgets why it downloaded a package

### ISSUE THREAD

**Issue by jyn514** [rust-lang/cargo#9222]
**Problem**

Cargo can download a package, then break when trying to parse its manifest. This makes it impossible to figure out why the package was downloaded.

**Steps**

1. git clone https://github.com/rust-lang/docs.rs/ && cd docs.rs
2. cargo update -Z minimal-versions
3. cargo tree -i phantom

```
error: failed to download `phantom v0.0.1`

Caused by:
  unable to get packages from source

Caused by:
  failed to parse manifest at `/home/joshua/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/phantom-0.0.1/Cargo.toml`

Caused by:
  invalid type: map, expected a string for key `lib`
```

**Possible Solution(s)**

- Don't make the error fatal
- Don't try to parse phantom's manifest in the first place; it shouldn't be necessary to figure out why the crate is needed.

**Notes**

Output of `cargo version`: cargo 1.52.0-nightly (bf5a5d5e5 2021-02-18)

**Comment by jyn514** [comment]
Note that `cargo update -p phantom:0.0.1` still works fine, so I don't think cargo inherently needs to give an error here.

**Comment by ehuss** [comment]
Most commands (`cargo tree` included) need to parse and read the manifest in order to fully understand the structure of a package.  In the case of `cargo tree`, there is not enough information in the index for it to do its job (mostly around proc-macros).  Unfortunately, this probably cannot be changed, at least not in any easy way.  

I realize it is not a very satisfactory answer, but the best workaround is to just look at the `Cargo.lock` file.

The error about "failed to download" is a bit confusing in this circumstance, since it did download, it just couldn't read the result.

**Comment by jyn514** [comment]
> Most commands (cargo tree included) need to parse and read the manifest in order to fully understand the structure of a package. 

Right, I understand that. My question is why it needs to read the manifest of *phantom*, and not just the manifest of the packages that depend on it. Basically I want to find what logic `cargo update` used to decide it needed phantom in the first place, which it can't have done with the manifest because it wasn't downloaded at the time.

**Comment by adamnovak** [comment]
I'm running into this problem because I'm trying to build a `2018` edition Rust package, with a Cargo that understands the `2018` edition but not the `2021` edition, and I need to be able to work out where in the dependency tree `2021` edition code is being depended on.

I can't do it with `cargo tree`, because `cargo tree` demands to parse the too-new package's manifest before it can explain why it wanted it.

This makes it impossible to find the package that is letting dependencies newer than its own edition into the tree.

**Comment by epage** [comment]
#15834 would remove the need for parsing manifests so we could complete the resolve.  We may still need additional information, depending on the flags (like `--format {l}`) but that seems like it would be a lot easier to isolate or make recoverable.
